### PR TITLE
Storage Blobs API

### LINKED ISSUE #1796: Blob Upload Support
This tracks the [Storage] Blob Upload/Download Support - Embedded-C for the Azure SDK team

### LINKED ISSUE #1797: Blob Download Support
Add support for blob download.

### PR DESCRIPTION
**Pull request by antkmsft**
Closes #1796.
Closes #1797.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -156,6 +156,7 @@ make ${project_name}_cov_xml //i.e. az_core_cov_xml or az_iot_cov_xml
 ## Code Coverage is available for these projects:
 #  az_core
 #  az_iot
+#  az_storage_blobs
```
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply CONTRIBUTING.md:159]: Pretty good line coverage. We probably want to track improving branch coverage.

https://dev.azure.com/azure-sdk/public/_build/results?buildId=1175218&_a=summary&tab=ms.vss-codecoverage-web.code-coverage-details&view=codecoverage-tab

Line coverage: 97.7% (86 of 88)
Branch coverage: 55.5% (40 of 72)

File: README.md [hunk 2]
```diff
@@ -65,6 +65,7 @@ To get help with the SDK:
 The Azure SDK for Embedded C repo has been structured around the service libraries it provides:
 
 1. [IoT](https://github.com/Azure/azure-sdk-for-c/blob/main/sdk/docs/iot) - Library to connect Embedded Devices to Azure IoT services
+2. [Storage](https://github.com/Azure/azure-sdk-for-c/blob/main/sdk/docs/storage) - Library to handle blob uploads to Azure Storage services
```
Comment (ahsonkhan) [inline README.md:68]:
```suggestion
2. [Storage](https://github.com/Azure/azure-sdk-for-c/blob/main/sdk/docs/storage) - Library to handle blob download from and upload to Azure Storage
```

File: samples/README.md [hunk 3]
```diff
@@ -34,4 +34,9 @@ This section lists how-to guides for the most commonly used APIs and most common
 
 ### Azure.IoT
 
-- [IoT samples](https://github.com/Azure/azure-sdk-for-c/blob/main/sdk/samples/iot/README.md)
+- [IoT samples](https://github.com/Azure/azure-sdk-for-c/blob/master/sdk/samples/iot/README.md)
+
+
```
Comment (ahsonkhan) [inline samples/README.md:39]: nit: remove extra new line
```suggestion
```

File: eng/pipelines/templates/stages/archetype-sdk-client.yml [hunk 4]
```diff
@@ -21,6 +21,9 @@ stages:
                 NonWindowsPath: ./build/sdk/samples/iot/paho_iot_hub_sas_telemetry_sample
                 WindowsPath: .\build\sdk\samples\iot\Debug\paho_iot_hub_sas_telemetry_sample.exe
                 RunCondition: contains(variables['build.args'], '-DTRANSPORT_PAHO=ON')
+              - Name: Storage Blobs
+                NonWindowsPath: ./build/sdk/samples/storage/blobs/blobs_client_sample
+                WindowsPath: build\sdk\samples\storage\blobs\debug\blobs_client_sample.exe
```
Comment (ahsonkhan) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:26]: Can we try triggering and running the storage blobs live sample tests in this PR to verify this change works as expected?

cc @danieljurek

File: sdk/docs/storage/LICENSE [hunk 5]
```diff
@@ -0,0 +1,21 @@
+    MIT License
```
Comment (ahsonkhan) [inline sdk/docs/storage/LICENSE:1]: @danewalton looks like we are missing the LICENSE file for `docs/iot/`
Maybe we should add it.

We have one at the root of the repo already, not sure what the protocol is. @RickWinter or @weshaggard do you know?

**Review by ahsonkhan — COMMENTED** [review]
Partial review, will continue tomorrow :)

File: sdk/samples/storage/blobs/src/blobs_client_sample.c [hunk 6]
```diff
@@ -68,18 +60,27 @@ int main()
...
 #endif
 
-  // enable logging
-  az_log_set_message_callback(write_log_message);
+  // Uncomment the line below to enable logging
+  // az_log_set_message_callback(write_log_message);
 
   // 1) Init client.
   // Example expects AZURE_BLOB_URL_WITH_SAS in env to be a URL w/ SAS token
   az_storage_blobs_blob_client client = { 0 };
```
Comment (JeffreyRichter) [inline sdk/samples/storage/blobs/src/blobs_client_sample.c:68]: I'd move this code to right before the _init call

**Review by JeffreyRichter — APPROVED** [review]

File: sdk/docs/storage/LICENSE (hunk 5 above)
Comment (weshaggard) [reply sdk/docs/storage/LICENSE:1]: In general I prefer to keep one license file per-repo. However in some languages the LICENSE file must live next to the source in the repo and be part of the package. For c given we ship as one unit I would stick with a single one in the root.

Comment (antkmsft) [reply sdk/docs/storage/LICENSE:1]: Thank you Wes, I'll remove the duplicate license file.

Comment (antkmsft) [reply sdk/docs/storage/LICENSE:1]: #1981

File: eng/pipelines/templates/stages/archetype-sdk-client.yml (hunk 4 above)
Comment (antkmsft) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:26]: I ran the sample locally, looks good.

**Pull request merged**